### PR TITLE
PP-2423 Handle ePDQ refund-related notifications

### PR DESCRIPTION
**Pull request by alexbishop1**
Map `7` (payment deleted) and `8` (refund) to `REFUNDED`. Map `73` (payment deletion refused), `83` (refund refused) and `94` (refund declined by the acquirer) to `REFUND ERROR`.

Change refund lookups to look up by the provider and the reference (not just the reference) to avoid potentially finding the wrong refund if providers have overlapping references.

with @juanjoqmelian 
with @SandorArpa

### CODE REVIEW

File: src/test/java/uk/gov/pay/connector/it/resources/worldpay/WorldpayNotificationResourceITest.java [hunk 1]
```diff
@@ -127,10 +148,18 @@ public void shouldFailWhenUnexpectedContentType() throws Exception {
...
+    }
+
+    private ValidatableResponse notifyConnector(String payload) throws Exception {
         String validIp = new DnsUtils().dnsLookup("build.ci.pymnt.uk").get();
         String xForwardedForHeader = format("%s, %s", validIp, "8.8.8.8");
         return given().port(app.getLocalPort())
-                .body(notificationPayloadForTransaction(transactionId, status))
+                .body(payload)
                 .header("X-Forwarded-For", xForwardedForHeader)
                 .contentType(TEXT_XML)
```
Comment (heathd) [inline src/test/java/uk/gov/pay/connector/it/resources/worldpay/WorldpayNotificationResourceITest.java:164]: is this correct? I thought ePDQ notifications were `application/x-www-form-urlencoded` encoded?

Comment (heathd) [reply src/test/java/uk/gov/pay/connector/it/resources/worldpay/WorldpayNotificationResourceITest.java:164]: oh, sorry this is the Worldpay test .. doh!

File: src/main/java/uk/gov/pay/connector/dao/RefundDao.java [hunk 2]
```diff
@@ -20,14 +20,17 @@ public RefundDao(final Provider<EntityManager> entityManager) {
...
 
-    public Optional<RefundEntity> findByReference(String reference) {
+    public Optional<RefundEntity> findByProviderAndReference(String provider, String reference) {
 
-        String query = "SELECT r FROM RefundEntity r " +
-                "WHERE r.reference = :reference";
+        String query = "SELECT refund FROM RefundEntity refund " +
+                "JOIN ChargeEntity charge ON refund.chargeEntity.id = charge.id " +
+                "JOIN GatewayAccountEntity gatewayAccount ON charge.gatewayAccount.id = gatewayAccount.id " +
+                "WHERE refund.reference = :reference AND gatewayAccount.gatewayName = :provider";
```
Comment (heathd) [inline src/main/java/uk/gov/pay/connector/dao/RefundDao.java:28]: can you check that there are suitable indexes on DB columns used in joins and where clauses?

Comment (alexbishop1) [reply src/main/java/uk/gov/pay/connector/dao/RefundDao.java:28]: The ID-related ones had indices but `reference` on `refunds` and `payment_provider` on `gateway_accounts` did not. We have added gubbins to `migrations.xml` to create indices for these.

Comment (SandorArpa) [reply src/main/java/uk/gov/pay/connector/dao/RefundDao.java:28]: Added relevant indexes (refunds.reference and gateway_accounts.payment_provider)

Comment (ckalista) [reply src/main/java/uk/gov/pay/connector/dao/RefundDao.java:28]: 👍  Especially now that we are supporting refund for more than just worldpay.

File: src/main/java/uk/gov/pay/connector/service/epdq/EpdqNotification.java [hunk 3]
```diff
@@ -14,69 +15,76 @@
...
+
+    public String getPayIdSub() {
+        return payIdSub;
+    }
 
-  public String getPayId() {
-    return payId;
-  }
+    public String getReference() {
+        if (Strings.isNullOrEmpty(payId) || Strings.isNullOrEmpty(payIdSub)) {
```
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/service/epdq/EpdqNotification.java:66]: Good to handle this in this way 👍

**Review by ckalista — COMMENTED** [review]
Looks good to me.

**Review by daniele-occhipinti — APPROVED** [review]
Looks very good to me.
Great you divided this big chunk of work in separate commits - that made the review far better. Thanks!

**Pull request merged**